ros/src/twist_controller/twist_controller.py

```diff
@@ -1,14 +1,73 @@
+from yaw_controller import YawController
+from pid import PID
+import rospy
+
+from lowpass import LowPassFilter #TODO: find out how to use it
 
 GAS_DENSITY = 2.858
 ONE_MPH = 0.44704
 
+ENABLE_BRAKES = True #Experment Brakes toggle
 
 class Controller(object):
     def __init__(self, *args, **kwargs):
         # TODO: Implement
+        '''args = [0:wheel_base,    1:steer_ratio,
+                   2:max_lat_accel, 3:max_steer_angle,
+                   4:decel_limit,   5:accel_limit
+                   6:brake_deadband,7:vehicle_mass ,
+                   8:fuel_capacity, 9:wheel_radius ,
+                   10:sample_rate
+        ]'''
+        
+        self.brake_deadband = args[6]
+        self.vehicle_mass   = args[7]
+        self.fuel_capacity  = args[8]
+        self.wheel_radius = args[9]
+
+        self.min_speed = (rospy.get_param('/waypoint_loader/velocity') * 1000.) / (60. * 60.) #kmph to mps
+        self.yaw_controller = YawController(args[0],#wheel_base
+                                            args[1],#steer_ratio
+                                            self.min_speed ,
+                                            args[2],#max_lat_accel
+                                            args[3])#max_steer_angle
+        self.sample_time = 1.0/args[6] # 1/sample_rate
+        self.lowpass_tau = 1#1 is default value, it should be the max steering value allowed to avoid jerk
+        self.lowpass_steer = LowPassFilter(1, self.sample_time)
+        self.pid_throttle = PID( 3.0, 0.0, 0.5, args[4], args[5] )
+        
         pass
 
     def control(self, *args, **kwargs):
         # TODO: Change the arg, kwarg list to suit your needs
-        # Return throttle, brake, steer
-        return 1., 0., 0.
+        ''' args = [0:proposed_linear_velocity,
+                    1:proposed_angular_velocity,
+                    2:current_linear_velocity,  
+                    3:current_angular_velocity
+                    4:is_dbw_enabled] from DBWNode dbw_node.py
+        '''
+        if args[4] : #if is_dbw_enabled
+            steer =  self.yaw_controller.get_steering(args[0],args[1],args[2])#bad and delayed but smooth
+            #steer = self.lowpass_steer.filt(steer)#TODO: test before uncomment and commit
+            throttle_CTE = args[0]-args[2] #proposed_linear_velocity - current_linear_velocity
+            throttle = self.pid_throttle.step(throttle_CTE,self.sample_time)#1/15 or 1/50
+            brake = 0
+            #Brakes system
+            if ENABLE_BRAKES : 
+                if throttle < self.brake_deadband*-1:#less than (0 - brake_deadband)
```

Comment:
Does it make sense to compare `throttle` with `brake_deadband` ?

Reply:
then make it less than 0 ?
I'm trying to ignore some less than 0 throttles , because car -seems- can't steer while braking

Reply:
I think throttle and breaks use different units, so comparing them seems incorrect. Comparing with `0` makes more sense to me.